src/webgpu/shader/execution/builtin/floor.spec.ts

```diff
@@ -0,0 +1,123 @@
+export const description = `
+Execution Tests for the 'floor' builtin function
+`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { GPUTest } from '../../../gpu_test.js';
+import { f32, f32Bits, TypeF32 } from '../../../util/conversion.js';
+
+import { anyOf, kBit, kValue, run } from './builtin.js';
+
+export const g = makeTestGroup(GPUTest);
+
+g.test('float_builtin_functions,floor')
+  .uniqueId('2edb4534aa0e48a6')
+  .specURL('https://www.w3.org/TR/2021/WD-WGSL-20210929/#float-builtin-functions')
+  .desc(
+    `
+floor:
+T is f32 or vecN<f32> floor(e: T ) -> T Returns the floor of e. Component-wise when T is a vector. (GLSLstd450Floor)
+`
+  )
+  .params(u =>
+    u
+      .combine('storageClass', ['uniform', 'storage_r', 'storage_rw'] as const)
+      .combine('vectorize', [undefined, 2, 3, 4] as const)
+  )
+  .fn(async t => {
+    run(t, 'floor', [TypeF32], TypeF32, t.params, [
+      // Small positive numbers
+      { input: f32(0.1), expected: f32(0.0) },
+      { input: f32(0.9), expected: f32(0.0) },
+      { input: f32(1.0), expected: f32(1.0) },
+      { input: f32(1.1), expected: f32(1.0) },
+      { input: f32(1.9), expected: f32(1.0) },
+
+      // Small negative numbers
+      { input: f32(-0.1), expected: f32(-1.0) },
+      { input: f32(-0.9), expected: f32(-1.0) },
+      { input: f32(-1.0), expected: f32(-1.0) },
+      { input: f32(-1.1), expected: f32(-2.0) },
+      { input: f32(-1.9), expected: f32(-2.0) },
+
+      // Min and Max f32
+      { input: f32Bits(kBit.f32.negative.max), expected: f32(-1.0) },
+      { input: f32Bits(kBit.f32.negative.min), expected: f32Bits(kBit.f32.negative.min) },
+      { input: f32Bits(kBit.f32.positive.min), expected: f32(0.0) },
+      { input: f32Bits(kBit.f32.positive.max), expected: f32Bits(kBit.f32.positive.max) },
```

Comment:
Is this right?

Reply:
I think these two are correct:
```
kBit.f32.negative.max is 0x80800000 (-1.17549435082e-38), so rounds to down to -1
kBit.f32.positive.min is 0x00800000 (1.17549435082e-38), so rounds to down to 0
```

I am not sure about the other two, since they are one the edge of what can be stored in a float:
```
kBit.f32.negative.min is 0xFF7FFFFF (-3.40282346639e38), so rounds down to -4e38, which isn't representable as a f32
kBit.f32.positive.max is 0x7F7FFFFF (3.40282346639e38) so should round to 3e38 which is something more like 0x7F61B1E6
```

I guess the question is floor meaningfully defined in these cases?

The actual text from the spec:
```
The floor expression is defined over real numbers x:

⌊x⌋ = k, where k is the unique integer such that k ≤ x < k+1
```

We have a similar mirror issue for ceil, which these tests were based on

Reply:
Okay, sorry, at these extremes there's no fractional at all. LGTM.